### ISSUE TITLE
Inconsistency due to local recursive functions

### ISSUE THREAD

**Issue by ia0** [FStarLang/FStar#1241]
In some circumstances (see example below), using a local recursive function (instead of a lifted top-level one) permits to prove false:
```
module Bug

open FStar.Seq

// This theoretically equivalent version does not prove false.
// let rec aux (n: pos) (i: nat { i <= n }):
//     Tot (r: seq bool { length r = n - i }) (decreases (n - i)) =
//   if i = n then createEmpty else append (create 1 true) (aux n (i + 1))
// let impl (n: pos): Tot (r: seq bool { length r = n }) = aux n 0

let impl (n: pos): Tot (r: seq bool { length r = n }) =
  let rec aux (i: nat { i <= n }):
      Tot (r: seq bool { length r = n - i }) (decreases (n - i)) =
    if i = n then createEmpty else append (create 1 true) (aux (i + 1))
  in aux 0

let foo: seq bool = impl 1
let bar: seq bool = impl 2
let () = assert false
```
This is maybe related to #380 but I could not reduce the example more to understand the root cause.

**Comment by aseemr** [comment]
My guess of what's happening: F*'s encoding of inner `let rec` is not precise. So, for the example above, the `aux` function is not encoded to the solver precisely. Instead, the encoding is:

```
(declare-fun let-rec_3309 () Term)
;;;;;;;;;;;;;;;;Equation for Test.impl
;;; Fact-ids: Name Test.impl; Namespace Test
(assert (! (forall ((@x0 Term))
		   (! (= (Test.impl @x0)
			 let-rec_3309)
```

Essentially saying, `impl x0 = let-rec_3309`, which means Z3 infers `foo = bar`, and from the refinement on the return type of `impl`, `0 = 1`. (If we change the refinement on the return type of `impl` to `length r <= n`, then assertion fails.)

The imprecise encoding of inner `let rec` needs to be fixed.

**Comment by aseemr** [comment]
I wonder if we could just omit the encoding of inner `let rec` altogether? The bit above seems not so useful anyway.

**Comment by nikswamy** [comment]
@aseemr  any news about this one? sounded like you were close to a fix yesterday

**Comment by aseemr** [comment]
@nikswamy, i plan to fix it today.

**Comment by aseemr** [comment]
Fixed in the commit 0104265fad1405e9a46a2682a82e8dea3d5025e2.

Thanks @ia0 for the report!
